libraries/AP_Scripting/examples/sub_test_synthetic_seafloor.lua

```diff
@@ -0,0 +1,701 @@
+    end
+
+    -- Check for delaying the measurement
+    if config.delay_s ~= 0.0 then
+        func = delay_funcfactory(config.delay_s, config.callback_interval_ms, func)
+    end
+
+    if func == nil then
+        func = identity_funcfactory()
+    end
+
+    return func
+end
+
+-------------------------------------------------------------------------------
+
+-- Range Model
+
+---@class RangeModel
+---@field get_range function(RangeModel, location_ud): number
+---@field sub_z_m number
+---@field bottom_z_m number
+---@field range_m number
+---@field set_origin function(RangeModel, location_ud)
+---@field is_origin_valid function(RangeModel)
+
+---@return RangeModel
+local function range_model_factory(model_bearing_N_rad, model_depth_m, vertices)
+
+    ---@type Location_ud
+    local origin_loc
+
+    local profile = profile_factory(vertices)
+  
+    ---@param sub_loc Location_ud
+    local function get_range(self, sub_loc)
+
+        -- Figure out the depth of the sub in absolute frame
+        sub_loc:change_alt_frame(0)
+        self.sub_z_m = -sub_loc:alt()/100
+
+        -- If the origin has not been set then do not use the profile
+        if not origin_loc then
+            self.bottom_z_m = model_depth_m
+            self.range_m = model_depth_m - self.sub_z_m
+            return self.range_m
+        end
+
+        -- The model origin has been set so use the profile.
+        -- N in name means relative to Earth frame, M in name means relative to Sea floor model frame.
+        local sub_bearing_N_rad = origin_loc:get_bearing(sub_loc)
+        local sub_distance_M_m = origin_loc:get_distance(sub_loc)
+
+        local sub_bearing_M_rad = sub_bearing_N_rad - model_bearing_N_rad
+        local sub_northly_M_m = math.cos(sub_bearing_M_rad) * sub_distance_M_m
+
+        -- profile:intersect has an origin at zero so we must calculate how far 
+        -- the sub is above that origin. And also calculate the depth of the
+        -- sea floor below the sub's location.
+        -- For the sea floor depth:
+        -- bottom_z = model_depth_m - profile(0)
+        -- Positive z is down, positive profile is up
+        local bottom_ray = ray_factory(sub_northly_M_m, 0, 0)
+        self.bottom_z_m = model_depth_m - profile:intersect(bottom_ray)
+
+        -- For the sub range measurement:
+        -- model_range = model_depth_m - sub_z
+        -- range = model_depth_m + profile(0) - sub_z;
+        -- range = profile(-model_range);
+        local sub_above_floor_m = model_depth_m - self.sub_z_m
+        local sub_ray = ray_factory(sub_northly_M_m, -sub_above_floor_m, 0)
+        self.range_m = profile:intersect(sub_ray)
+        return self.range_m
+    end
+
+    return {
+        get_range = get_range,
+        sub_z_m = 0,
+        bottom_z_m = model_depth_m,
+        range_m = model_depth_m,
+        set_origin = function(self, origin) origin_loc = origin end,
+        is_origin_valid = function(self) return origin_loc ~= nil end,
```

Comment:
The CLion Lua plugin told me that `self` was unused and that `origin_loc` was a global symbol, which caused me to learn about closures and the syntactic sugar around calling functions with ":". Interesting!

I suppose the CLion plugin shouldn't complain about this, but you could change the reference to `self.origin_loc`, then the plugin is happy as-is and it is more obvious to a Lua neophyte like myself.